lib/kernel.js

```diff
@@ -74,31 +99,58 @@ export default class Kernel {
   }
 
   interrupt() {
-    throw new Error("Kernel: interrupt method not implemented");
+    this.rawInterrupt();
   }
 
-  shutdown() {
-    throw new Error("Kernel: shutdown method not implemented");
+  shutdown(restart: ?boolean, onRestarted: ?Function) {
+    if (restart === true) {
+      this.middleware.shutdown(restart, onRestarted);
+    } else {
+      this.middleware.shutdown(false);
```

Comment:
I'd prefer to keep shutdown and restart separate methods. The restart argument is coming from the [jupyter messaging protocol](http://jupyter-client.readthedocs.io/en/latest/messaging.html#kernel-shutdown) and doesn't actually restart the kernel.

Reply:
I see that there's a restart argument in the messaging protocol, but does that argument even serve a useful purpose that hydrogen cares about?

If not, maybe it would still make sense to combine restart and shutdown in the middleware? By this I mean that `restart: ?boolean` will differentiate between true restart/shutdown rather than controlling the messaging protocol argument.

Reply:
Actually what I said above is already the case in the code: setting restart=true here will initiate a true restart, not just toggle a flag in the shutdown message.

I'm not sure that exposing this restart flag to plugins would be useful in any way. (That said, the case can be made that separating shutdown/restart middleware will avoid needless confusion.)